Crash in WebCore::PlaybackSessionModelMediaElement::exitFullscreen due to null m_mediaElement

aestes opened the pull request:
#### e56299b4f1d9bb84f69ad5cb7ee24fa6fc93eb56
<pre>
Crash in WebCore::PlaybackSessionModelMediaElement::exitFullscreen due to null m_mediaElement
<a href="https://bugs.webkit.org/show_bug.cgi?id=276341">https://bugs.webkit.org/show_bug.cgi?id=276341</a>
<a href="https://rdar.apple.com/131143888">rdar://131143888</a>

Reviewed by Jean-Yves Avenard and Eric Carlson.

Despite an ASSERT(element) in PlaybackSessionModelMediaElement::exitFullscreen, crash reports
indicate that m_mediaElement can be null when exitFullscreen is called. Resolved this by
null-checking m_mediaElement prior to dereferencing it (in exitFullscreen and elsewhere); the debug
assertion is left to help investigate why m_mediaElement is unexpectedly null. While here, updated
PlaybackSessionModelMediaElement to follow our latest smart pointer usage guidelines.

* Source/WebCore/platform/cocoa/PlaybackSessionModelMediaElement.mm:
(WebCore::PlaybackSessionModelMediaElement::setMediaElement):
(WebCore::PlaybackSessionModelMediaElement::play):
(WebCore::PlaybackSessionModelMediaElement::pause):
(WebCore::PlaybackSessionModelMediaElement::togglePlayState):
(WebCore::PlaybackSessionModelMediaElement::beginScrubbing):
(WebCore::PlaybackSessionModelMediaElement::endScrubbing):
(WebCore::PlaybackSessionModelMediaElement::seekToTime):
(WebCore::PlaybackSessionModelMediaElement::fastSeek):
(WebCore::PlaybackSessionModelMediaElement::beginScanningForward):
(WebCore::PlaybackSessionModelMediaElement::beginScanningBackward):
(WebCore::PlaybackSessionModelMediaElement::endScanning):
(WebCore::PlaybackSessionModelMediaElement::setDefaultPlaybackRate):
(WebCore::PlaybackSessionModelMediaElement::setPlaybackRate):
(WebCore::PlaybackSessionModelMediaElement::selectAudioMediaOption):
(WebCore::PlaybackSessionModelMediaElement::selectLegibleMediaOption):
(WebCore::PlaybackSessionModelMediaElement::togglePictureInPicture):
(WebCore::PlaybackSessionModelMediaElement::toggleInWindowFullscreen):
(WebCore::PlaybackSessionModelMediaElement::enterFullscreen):
(WebCore::PlaybackSessionModelMediaElement::exitFullscreen):
(WebCore::PlaybackSessionModelMediaElement::setMuted):
(WebCore::PlaybackSessionModelMediaElement::setVolume):
(WebCore::PlaybackSessionModelMediaElement::setPlayingOnSecondScreen):
(WebCore::PlaybackSessionModelMediaElement::spatialTrackingLabel const):
(WebCore::PlaybackSessionModelMediaElement::setSpatialTrackingLabel):
(WebCore::PlaybackSessionModelMediaElement::sendRemoteCommand):
(WebCore::PlaybackSessionModelMediaElement::updateMediaSelectionOptions):
(WebCore::PlaybackSessionModelMediaElement::playbackStartedTime const):
(WebCore::PlaybackSessionModelMediaElement::duration const):
(WebCore::PlaybackSessionModelMediaElement::currentTime const):
(WebCore::PlaybackSessionModelMediaElement::bufferedTime const):
(WebCore::PlaybackSessionModelMediaElement::isPlaying const):
(WebCore::PlaybackSessionModelMediaElement::isStalled const):
(WebCore::PlaybackSessionModelMediaElement::defaultPlaybackRate const):
(WebCore::PlaybackSessionModelMediaElement::playbackRate const):
(WebCore::PlaybackSessionModelMediaElement::seekableRanges const):
(WebCore::PlaybackSessionModelMediaElement::seekableTimeRangesLastModifiedTime const):
(WebCore::PlaybackSessionModelMediaElement::liveUpdateInterval const):
(WebCore::PlaybackSessionModelMediaElement::canPlayFastReverse const):
(WebCore::PlaybackSessionModelMediaElement::audioMediaSelectionOptions const):
(WebCore::PlaybackSessionModelMediaElement::legibleMediaSelectionOptions const):
(WebCore::PlaybackSessionModelMediaElement::legibleMediaSelectedIndex const):
(WebCore::PlaybackSessionModelMediaElement::externalPlaybackEnabled const):
(WebCore::PlaybackSessionModelMediaElement::externalPlaybackTargetType const):
(WebCore::PlaybackSessionModelMediaElement::externalPlaybackLocalizedDeviceName const):
(WebCore::PlaybackSessionModelMediaElement::wirelessVideoPlaybackDisabled const):
(WebCore::PlaybackSessionModelMediaElement::isMuted const):
(WebCore::PlaybackSessionModelMediaElement::volume const):
(WebCore::PlaybackSessionModelMediaElement::isPictureInPictureSupported const):
(WebCore::PlaybackSessionModelMediaElement::isPictureInPictureActive const):
(WebCore::PlaybackSessionModelMediaElement::isInWindowFullscreenActive const):
(WebCore::PlaybackSessionModelMediaElement::logIdentifier const):
(WebCore::PlaybackSessionModelMediaElement::loggerPtr const):

Canonical link: <a href="https://commits.webkit.org/280780@main">https://commits.webkit.org/280780@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/be8a4785e9f04f77175cff23b5e7a9115d63eaff

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/57563 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/36891 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/10038 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/61185 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/8008 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/59691 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/44524 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/8196 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/46607 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/5678 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/59593 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/34610 "Found 1 new test failure: compositing/reflections/load-video-in-reflection.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/49725 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/27473 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/31387 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/7025 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/7011 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/53346 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/7298 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/62865 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/1476 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/7389 "layout-tests (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/53867 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/1483 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/49747 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/53969 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12745 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/1255 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/32720 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/33805 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/34890 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/33551 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

jyavenard commented on Source/WebCore/platform/cocoa/PlaybackSessionModelMediaElement.mm:
```diff
@@ -59,8 +59,11 @@
+    RefPtr oldMediaElement = m_mediaElement;
+    RefPtr newMediaElement = mediaElement;
+
+    if (oldMediaElement.get() == newMediaElement.get()) {
```
.get() seems redundant here

jyavenard commented on Source/WebCore/platform/cocoa/PlaybackSessionModelMediaElement.mm:
```diff
@@ -337,13 +342,13 @@
 {
 #if ENABLE(VIDEO_PRESENTATION_MODE)
-    auto* element = dynamicDowncast<HTMLVideoElement>(*m_mediaElement);
+    RefPtr element = dynamicDowncast<HTMLVideoElement>(m_mediaElement);
```
it seems to me that only this line was required to fix the crashes. Do we need all the above ?

aestes replied:
No, but it seemed like a good opportunity to make these mechanical changes.

jyavenard commented on Source/WebCore/platform/cocoa/PlaybackSessionModelMediaElement.mm:
```diff
@@ -514,65 +521,90 @@
 Ref<TimeRanges> PlaybackSessionModelMediaElement::seekableRanges() const
 {
-    return m_mediaElement && m_mediaElement->supportsSeeking() ? m_mediaElement->seekable() : TimeRanges::create();
+    if (RefPtr mediaElement = m_mediaElement) {
```
this could be done more concisely as:
```
if (RefPtr mediaElement = m_mediaElement; mediaElement && mediaElement->supportsSeeking())
    return mediaElement->seekable();
```

jyavenard reviewed (approved):
r=me with nits

aestes committed:
Crash in WebCore::PlaybackSessionModelMediaElement::exitFullscreen due to null m_mediaElement
https://bugs.webkit.org/show_bug.cgi?id=276341
rdar://131143888

Reviewed by Jean-Yves Avenard and Eric Carlson.

Despite an ASSERT(element) in PlaybackSessionModelMediaElement::exitFullscreen, crash reports
indicate that m_mediaElement can be null when exitFullscreen is called. Resolved this by
null-checking m_mediaElement prior to dereferencing it (in exitFullscreen and elsewhere); the debug
assertion is left to help investigate why m_mediaElement is unexpectedly null. While here, updated
PlaybackSessionModelMediaElement to follow our latest smart pointer usage guidelines.

* Source/WebCore/platform/cocoa/PlaybackSessionModelMediaElement.mm:
(WebCore::PlaybackSessionModelMediaElement::setMediaElement):
(WebCore::PlaybackSessionModelMediaElement::play):
(WebCore::PlaybackSessionModelMediaElement::pause):
(WebCore::PlaybackSessionModelMediaElement::togglePlayState):
(WebCore::PlaybackSessionModelMediaElement::beginScrubbing):
(WebCore::PlaybackSessionModelMediaElement::endScrubbing):
(WebCore::PlaybackSessionModelMediaElement::seekToTime):
(WebCore::PlaybackSessionModelMediaElement::fastSeek):
(WebCore::PlaybackSessionModelMediaElement::beginScanningForward):
(WebCore::PlaybackSessionModelMediaElement::beginScanningBackward):
(WebCore::PlaybackSessionModelMediaElement::endScanning):
(WebCore::PlaybackSessionModelMediaElement::setDefaultPlaybackRate):
(WebCore::PlaybackSessionModelMediaElement::setPlaybackRate):
(WebCore::PlaybackSessionModelMediaElement::selectAudioMediaOption):
(WebCore::PlaybackSessionModelMediaElement::selectLegibleMediaOption):
(WebCore::PlaybackSessionModelMediaElement::togglePictureInPicture):
(WebCore::PlaybackSessionModelMediaElement::toggleInWindowFullscreen):
(WebCore::PlaybackSessionModelMediaElement::enterFullscreen):
(WebCore::PlaybackSessionModelMediaElement::exitFullscreen):
(WebCore::PlaybackSessionModelMediaElement::setMuted):
(WebCore::PlaybackSessionModelMediaElement::setVolume):
(WebCore::PlaybackSessionModelMediaElement::setPlayingOnSecondScreen):
(WebCore::PlaybackSessionModelMediaElement::spatialTrackingLabel const):
(WebCore::PlaybackSessionModelMediaElement::setSpatialTrackingLabel):
(WebCore::PlaybackSessionModelMediaElement::sendRemoteCommand):
(WebCore::PlaybackSessionModelMediaElement::updateMediaSelectionOptions):
(WebCore::PlaybackSessionModelMediaElement::playbackStartedTime const):
(WebCore::PlaybackSessionModelMediaElement::duration const):
(WebCore::PlaybackSessionModelMediaElement::currentTime const):
(WebCore::PlaybackSessionModelMediaElement::bufferedTime const):
(WebCore::PlaybackSessionModelMediaElement::isPlaying const):
(WebCore::PlaybackSessionModelMediaElement::isStalled const):
(WebCore::PlaybackSessionModelMediaElement::defaultPlaybackRate const):
(WebCore::PlaybackSessionModelMediaElement::playbackRate const):
(WebCore::PlaybackSessionModelMediaElement::seekableRanges const):
(WebCore::PlaybackSessionModelMediaElement::seekableTimeRangesLastModifiedTime const):
(WebCore::PlaybackSessionModelMediaElement::liveUpdateInterval const):
(WebCore::PlaybackSessionModelMediaElement::canPlayFastReverse const):
(WebCore::PlaybackSessionModelMediaElement::audioMediaSelectionOptions const):
(WebCore::PlaybackSessionModelMediaElement::legibleMediaSelectionOptions const):
(WebCore::PlaybackSessionModelMediaElement::legibleMediaSelectedIndex const):
(WebCore::PlaybackSessionModelMediaElement::externalPlaybackEnabled const):
(WebCore::PlaybackSessionModelMediaElement::externalPlaybackTargetType const):
(WebCore::PlaybackSessionModelMediaElement::externalPlaybackLocalizedDeviceName const):
(WebCore::PlaybackSessionModelMediaElement::wirelessVideoPlaybackDisabled const):
(WebCore::PlaybackSessionModelMediaElement::isMuted const):
(WebCore::PlaybackSessionModelMediaElement::volume const):
(WebCore::PlaybackSessionModelMediaElement::isPictureInPictureSupported const):
(WebCore::PlaybackSessionModelMediaElement::isPictureInPictureActive const):
(WebCore::PlaybackSessionModelMediaElement::isInWindowFullscreenActive const):
(WebCore::PlaybackSessionModelMediaElement::logIdentifier const):
(WebCore::PlaybackSessionModelMediaElement::loggerPtr const):

Canonical link: https://commits.webkit.org/280780@main
